validator.go

```diff
@@ -0,0 +1,261 @@
+package jwt
+
+import (
+	"crypto/subtle"
+	"time"
+)
+
+// validator is the core of the new Validation API. It is automatically used by
+// a [Parser] during parsing and can be modified with various parser options.
+//
+// Note: This struct is intentionally not exported (yet) as we want to
+// internally finalize its API. In the future, we might make it publicly available.
+type validator struct {
+	// leeway is an optional leeway that can be provided to account for clock skew.
+	leeway time.Duration
+
+	// timeFunc is used to supply the current time that is needed for
+	// validation. If unspecified, this defaults to time.Now.
+	timeFunc func() time.Time
+
+	// verifyIat specifies whether the iat (Issued At) claim will be verified.
+	// According to https://www.rfc-editor.org/rfc/rfc7519#section-4.1.6 this
+	// only specifies the age of the token, but no validation check is
+	// necessary. However, if wanted, it can be checked if the iat is
+	// unrealistic, i.e., in the future.
+	verifyIat bool
+
+	// expectedAud contains the audience this token expects. Supplying an empty
+	// string will disable aud checking.
+	expectedAud string
+
+	// expectedIss contains the issuer this token expects. Supplying an empty
+	// string will disable iss checking.
+	expectedIss string
+
+	// expectedSub contains the subject this token expects. Supplying an empty
+	// string will disable sub checking.
+	expectedSub string
+}
+
+// CustomClaims represents a custom claims interface, which can be built upon the integrated
+// claim types, such as map claims or registered claims.
+type CustomClaims interface {
+	// CustomValidation can be implemented by a user-specific claim to support
+	// additional validation steps in addition to the regular validation.
+	CustomValidation() error
+}
+
+// newValidator can be used to create a stand-alone validator with the supplied
+// options. This validator can then be used to validate already parsed claims.
+func newValidator(opts ...ParserOption) *validator {
+	p := NewParser(opts...)
+	return p.validator
+}
+
+// Validate validates the given claims. It will also perform any custom
+// validation if claims implements the CustomValidator interface.
+func (v *validator) Validate(claims Claims) error {
+	var now time.Time
+	vErr := new(ValidationError)
+
+	// Check, if we have a time func
+	if v.timeFunc != nil {
+		now = v.timeFunc()
+	} else {
+		now = time.Now()
+	}
+
+	// We always need to check the expiration time, but usage of the claim
+	// itself is OPTIONAL
+	if !v.VerifyExpiresAt(claims, now, false) {
+		vErr.Inner = ErrTokenExpired
+		vErr.Errors |= ValidationErrorExpired
```

Comment:
I know I asked about this before, but can we remove these operators in favor of plain ol' errors?

Reply:
Yes and no :) We could for example replace `Errors` with a slice or `error` instead of the bitfield. I could then adjust the `errors.Is` to work with that. Once we only support Go 1.20+ we could even make use of the new `errors.Join` function, and `Unwrap` of multiple errors (see https://tip.golang.org/doc/go1.20#errors).

I can try to play around with that in a separate PR that merges into the branch.

Reply:
I played around with the idea and this is what I came up with: https://github.com/golang-jwt/jwt/pull/274.

However that uses/needs Go 1.20 for the new fancy multiple error wrapping (which is actually *exactly* what we need). While we had this discussion before, I personally could live with the idea of the `v5` branch "starting" at Go 1.20. Otherwise we could probably have some legacy layer that replicates the behaviour and then switch to the native functions at some point.

Reply:
Thanks for taking a stab at this, after a bit more thought I suggest we leave this as-is (for now). The most important thing is users don't actually have to do the bit manipulation, and can use `errors` to assert for the correct errors.

I think once Go reaches v1.21 or higher we can revisit the errors story (hopefully it'll just be an implemntation detail).

Reply:
> Thanks for taking a stab at this, after a bit more thought I suggest we leave this as-is (for now). The most important thing is users don't actually have to do the bit manipulation, and can use `errors` to assert for the correct errors.

Yes that can. 

> 
> I think once Go reaches v1.21 or higher we can revisit the errors story (hopefully it'll just be an implemntation detail).

Although a new major version might be a good point to make such changes and I actually quite like the very clean approach it could offer in terms of errors.

I will see whether I can play around with "simulating" the Go 1.20 behaviour on Go < 1.20 by basically keeping an (unexported) `validationError` (which acts as a compatibility layer) and once we have +2 release versions, we can just go to Go 1.20 as min version and drop the compatibility layer.

Reply:
I presume we could have 2 branches of code guarded by build tags? Although this may be more effort than it's worth.

If it comes down to that, I can get behind the `/v5` version being >=v1.20 minimum version

Reply:
> I presume we could have 2 branches of code guarded by build tags? Although this may be more effort than it's worth.
> 
> If it comes down to that, I can get behind the `/v5` version being >=v1.20 minimum version

I basically just implemented it in less than an hour. Can you maybe have a look at that PR?

Reply:
Sure thing. I'll take a look at that there and mark this conversation resolved here.